Lib/warnings.py

```diff
@@ -590,17 +590,54 @@ def __call__(self, arg, /):
         if category is None:
             arg.__deprecated__ = msg
             return arg
-        elif isinstance(arg, type):
+
+        def update_signature(original_func):
+            # Ensure that the signature of the decorated callable matches the original one
+
+            def wrapper(func):
+                import inspect
+
+                try:
+                    original_signature = inspect.signature(original_func)
```

Comment:
On Python 3.14 this will potentially evaluate `original_func`'s annotations:

```
>>> import inspect
>>> inspect.signature(f)
Traceback (most recent call last):
...
  File "<python-input-0>", line 1, in __annotate__
    def f(x: a): pass
             ^
NameError: name 'a' is not defined
```

Why isn't the `functools.wraps` call enough to allow `inspect` to get the signature of the underlying function?

Reply:
> Why isn't the `functools.wraps` call enough to allow `inspect` to get the signature of the underlying function?

In general use cases, the wrapped function may not have the same signature as the original function. So it is not a good idea to populate the signature in `functools.wraps`.

```python
def normalize2(func):
    @functools.wraps(func):
    def wrapped(x, y, /):
        return func(x, y)

    return wrapped


@normalize2
def f2(a, b): ...

@normalize2
def g2(u, v): ...
```

Reply:
`inspect.signature` will unwrap decorated functions (arbitrary level) and then get the underlying functions' signature.

For classes (also a callable), their signature is calculated from `meta.__call__`, `cls.__new__`, and `cls.__init__` based on MRO.

https://github.com/python/cpython/blob/04bc681e7cf5b7effd99e676b7deceb27fb34434/Lib/inspect.py#L2505-L2563

The `@deprecated` decorator creates a `__new__` method which previously might not exist. `inspect` will run into the wrong underlying method to inspect with.